// contentful-management/dist/es-modules/adapters/REST/endpoints/environment-template.js
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
import copy from 'fast-copy';
import * as raw from './raw';
const apiPath = (organizationId, ...pathSegments) => `/organizations/${organizationId}/environment_templates/` + pathSegments.join('/');
export const get = (http, {
  organizationId,
  environmentTemplateId,
  version,
  query = {}
}, headers) => version ? raw.get(http, apiPath(organizationId, environmentTemplateId, 'versions', version), {
  params: query,
  headers
}) : raw.get(http, apiPath(organizationId, environmentTemplateId), {
  params: query,
  headers
});
export const getMany = (http, {
  organizationId,
  query = {}
}, headers) => raw.get(http, apiPath(organizationId), {
  params: query,
  headers
});
export const create = (http, {
  organizationId
}, payload, headers) => raw.post(http, apiPath(organizationId), payload, {
  headers
});
export const update = (http, {
  organizationId,
  environmentTemplateId
}, payload, headers) => {
  var _payload$sys$version;
  const data = copy(payload);
  delete data.sys;
  return raw.put(http, apiPath(organizationId, environmentTemplateId), data, {
    headers: _objectSpread({
      'X-Contentful-Version': (_payload$sys$version = payload.sys.version) !== null && _payload$sys$version !== void 0 ? _payload$sys$version : 0
    }, headers)
  });
};
export const versionUpdate = (http, {
  organizationId,
  version,
  environmentTemplateId
}, payload, headers) => raw.patch(http, apiPath(organizationId, environmentTemplateId, 'versions', version), payload, {
  headers
});
export const del = (http, {
  organizationId,
  environmentTemplateId
}, headers) => raw.del(http, apiPath(organizationId, environmentTemplateId), {
  headers
});
export const versions = (http, {
  organizationId,
  environmentTemplateId,
  query = {}
}, headers) => raw.get(http, apiPath(organizationId, environmentTemplateId, 'versions'), {
  params: query,
  headers
});
export const validate = (http, {
  spaceId,
  environmentId,
  environmentTemplateId,
  version
}, payload, headers) => raw.put(http, version ? `/spaces/${spaceId}/environments/${environmentId}/template_installations/${environmentTemplateId}/versions/${version}/validated` : `/spaces/${spaceId}/environments/${environmentId}/template_installations/${environmentTemplateId}/validated`, payload, {
  headers
});
export const install = (http, {
  spaceId,
  environmentId,
  environmentTemplateId
}, payload, headers) => raw.post(http, `/spaces/${spaceId}/environments/${environmentId}/template_installations/${environmentTemplateId}/versions`, payload, {
  headers
});
export const disconnect = (http, {
  spaceId,
  environmentId,
  environmentTemplateId
}, headers) => raw.del(http, `/spaces/${spaceId}/environments/${environmentId}/template_installations/${environmentTemplateId}`, {
  headers
});
