// @contentful/app-scripts/lib/upload/build-upload-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildAppUploadSettings = void 0;
const inquirer_1 = require("inquirer");
const get_app_info_1 = require("../get-app-info");
const utils_1 = require("../utils");
const constants_1 = require("../constants");
async function buildAppUploadSettings(options) {
    const actionsManifest = (0, utils_1.getEntityFromManifest)('actions');
    const deliveryFnManifest = (0, utils_1.getEntityFromManifest)('deliveryFunctions');
    const prompts = [];
    const { bundleDir, comment, skipActivation, host } = options;
    if (!bundleDir) {
        prompts.push({
            name: 'bundleDirectory',
            message: `Bundle directory, if not default:`,
            default: './build',
        });
    }
    if (!comment) {
        prompts.push({
            name: 'comment',
            message: `Add a comment to the created bundle:`,
            default: '',
        });
    }
    if (skipActivation === undefined) {
        prompts.push({
            type: 'confirm',
            name: 'activateBundle',
            message: `Do you want to activate the bundle after upload?`,
            default: true,
        });
    }
    if (!host) {
        prompts.push({
            name: 'host',
            message: `Contentful CMA endpoint URL:`,
            default: constants_1.DEFAULT_CONTENTFUL_API_HOST,
        });
    }
    const { activateBundle, ...appUploadSettings } = await (0, inquirer_1.prompt)(prompts);
    const appInfo = await (0, get_app_info_1.getAppInfo)(options);
    return {
        bundleDirectory: bundleDir,
        skipActivation: skipActivation === undefined ? !activateBundle : skipActivation,
        comment,
        host,
        actions: actionsManifest,
        deliveryFunctions: deliveryFnManifest,
        ...appUploadSettings,
        ...appInfo,
    };
}
exports.buildAppUploadSettings = buildAppUploadSettings;
