// @contentful/app-scripts/lib/activate/activate-bundle.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.activateBundle = void 0;
const ora_1 = __importDefault(require("ora"));
const chalk_1 = require("chalk");
const utils_1 = require("../utils");
const contentful_management_1 = require("contentful-management");
async function activateBundle({ accessToken, organization, definition, bundleId, host, }) {
    const activationSpinner = (0, ora_1.default)('Activating your bundle').start();
    const plainClient = (0, contentful_management_1.createClient)({ accessToken, host }, { type: 'plain' });
    const defaultLocations = [{ location: 'dialog' }];
    const currentDefinition = await plainClient.appDefinition.get({
        appDefinitionId: definition.value,
        organizationId: organization.value,
    });
    currentDefinition.bundle = {
        sys: {
            id: bundleId,
            linkType: 'AppBundle',
            type: 'Link',
        },
    };
    if (!currentDefinition.locations?.length) {
        currentDefinition.locations = defaultLocations;
    }
    delete currentDefinition.src;
    try {
        await plainClient.appDefinition.update({
            appDefinitionId: definition.value,
            organizationId: organization.value,
        }, currentDefinition);
    }
    catch (err) {
        (0, utils_1.throwError)(err, 'Something went wrong activating your bundle. Make sure you used the correct definition-id.');
    }
    finally {
        activationSpinner.stop();
    }
    console.log(`
  ${(0, chalk_1.cyan)('Success!')} Your app bundle was activated for ${(0, chalk_1.cyan)(definition.name)} in ${(0, chalk_1.bold)(organization.name)}.
  Bundle Id: ${(0, chalk_1.yellow)(bundleId)}`);
    console.log(`
  ----------------------------
  Ready to share your app with the world? Submit it to the Developer Showcase (${(0, chalk_1.cyan)('https://ctfl.io/dev-showcase')}) or our Marketplace (${(0, chalk_1.cyan)('https://ctfl.io/submit-app')}).
 `);
}
exports.activateBundle = activateBundle;
